### PR TITLE
[US3437] feat(mayactl): add pool list command in mayactl

### PR DESCRIPTION
**Pull request by ashishranjan738**
Signed-off-by: Ashish Ranjan <ashishranjan738@gmail.com>

**What this PR does / why we need it**:
-> Add pool list command in mayactl which will currentyl list all cstor storagepools.

**Special notes for your reviewer**:
Output of mayactl list: 
```
ashishranjan738@Ashish-PC:~/Projects/go/src/github.com/openebs/maya$ mayactl -m 10.0.0.81 pool list

POOL NAME                  NODE NAME          POOL TYPE
---------                  ---------          ---------     
cstor-sparse-pool-3tzn     127.0.0.1          striped
```

### CODE REVIEW

File: pkg/install/v1alpha1/storage_pool.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+apiVersion: openebs.io/v1alpha1
+kind: RunTask
+metadata:
+  name: storage-pool-read-default
+spec: 
+  meta: |
+    id: readstoragepool
+    apiVersion: openebs.io/v1alpha1
+    kind: StoragePool
+    action: list
```
Comment (AmitKumarDas) [inline pkg/install/v1alpha1/storage_pool.go]: this should be read

File: pkg/install/v1alpha1/storage_pool.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    options: |-
```
Comment (AmitKumarDas) [inline pkg/install/v1alpha1/storage_pool.go]: options will not be used for read.

File: pkg/storagepool/storagepool.go [hunk 3]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+
+	castName := menv.Get(menv.CASTemplateToListStoragePoolENVK)
+
+	// fetch read cas template specifications
+	cast, err := s.k8sClient.GetOEV1alpha1CAST(castName, mach_apis_meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	glog.Infof("%v", cast)
```
Comment (AmitKumarDas) [inline pkg/storagepool/storagepool.go]: remove this

File: pkg/storagepool/storagepool.go [hunk 4]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := engine.Run()
+	if err != nil {
+		return nil, err
+	}
+
+	glog.Infof("%s", data)
```
Comment (AmitKumarDas) [inline pkg/storagepool/storagepool.go]: remove this

File: cmd/maya-apiserver/app/server/http.go [hunk 5]
```diff
@@ -120,6 +120,17 @@ var (
 		[]string{"code", "method"},
 	)
 
+	latestOpenEBSPoolRequestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "latest_openebs_pool_request_duration_seconds",
+			Help:    "Request response time of the /latest/pool/.",
+			Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.5, 1, 2.5, 5, 10},
```
Comment (sonasingh46) [inline cmd/maya-apiserver/app/server/http.go]: I would prefer some comments here for better understanding. It helps new comers.

File: cmd/maya-apiserver/app/server/http.go [hunk 6]
```diff
@@ -285,6 +304,8 @@ func (s *HTTPServer) registerHandlers(serviceProvider string, enableDebug bool)
 	// request for metrics is handled here. It displays metrics related to
 	// garbage collection, process, cpu...etc, and the custom metrics created.
 	s.mux.Handle("/metrics", promhttp.Handler())
+
```
Comment (sonasingh46) [inline cmd/maya-apiserver/app/server/http.go]: Can we remove this extra line?

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package server
```
Comment (sonasingh46) [inline cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: Can we add the `Copyright ...` comment.

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	pool "github.com/openebs/maya/pkg/storagepool"
+)
+
+type poolAPIOpsV1alpha1 struct {
+	req  *http.Request
+	resp http.ResponseWriter
+}
+
+// volumeV1alpha1SpecificRequest is a http handler to handle HTTP
```
Comment (sonasingh46) [inline cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: Wrong function name in comment.

File: pkg/storagepool/storagepool.go [hunk 9]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+
+	// unmarshall into openebs CasPool
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: Comment contains some mistake.

File: pkg/storagepool/storagepool.go [hunk 10]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	sPool := &v1alpha1.StoragePool{}
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: sPool---> sp
or 
sPool ----> storagePool
Let take view of  @AmitKumarDas . I am not sure.

File: pkg/storagepool/storagepool.go [hunk 11]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	err = json.Unmarshal(data, sPool)
+	if err != nil {
+		return nil, err
+	}
+	return sPool, nil
+}
+
+func (s *storagePoolOperation) Read() (*v1alpha1.StoragePool, error) {
+	if s.k8sClient == nil {
+		return nil, fmt.Errorf("Unable to fetch K8s client")
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: While forming and returning an error first letter should be small.

File: pkg/storagepool/storagepool.go [hunk 12]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := engine.Run()
+	if err != nil {
+		return nil, err
+	}
+
+	// unmarshall into openebs CasPool
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: Incorrect comment.

File: pkg/storagepool/storagepool.go [hunk 13]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	// Put pool object inside casPoolOperation object
+	return &storagePoolOperation{
+		poolName:  poolName,
+		k8sClient: kc,
+	}, err
+}
+
+func (s *storagePoolOperation) List() (*v1alpha1.StoragePoolList, error) {
+	if s.k8sClient == nil {
+		return nil, fmt.Errorf("Unable to fetch K8s client")
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: First letter should be small in error msg

File: pkg/storagepool/storagepool.go [hunk 14]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
...
+	poolName  string
+	k8sClient *m_k8s_client.K8sClient
+}
+
+func NewStoragePoolOperation(poolName string) (*storagePoolOperation, error) {
+	kc, err := m_k8s_client.NewK8sClient("")
+	if err != nil {
+		return nil, err
+	}
+	// Put pool object inside casPoolOperation object
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: Incorrect comment

File: pkg/storagepool/storagepool.go [hunk 15]
```diff
@@ -162,3 +165,97 @@ func (v *casPoolOperation) Delete() (*v1alpha1.CasPool, error) {
 	}
 	return pool, nil
 }
+
+type storagePoolOperation struct {
+	poolName  string
+	k8sClient *m_k8s_client.K8sClient
+}
+
+func NewStoragePoolOperation(poolName string) (*storagePoolOperation, error) {
```
Comment (sonasingh46) [inline pkg/storagepool/storagepool.go]: This is an exported function and should have a comment.

File: pkg/install/v1alpha1/env_setter_test.go [hunk 16]
```diff
@@ -28,7 +28,7 @@ func TestEnvInstallCount(t *testing.T) {
 	tests := map[string]struct {
 		expectedCount int
 	}{
-		"101": {16},
+		"101": {18},
```
Comment (sonasingh46) [inline pkg/install/v1alpha1/env_setter_test.go]: A comment would help here.

File: pkg/client/mapiserver/pool.go [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+	"encoding/json"
+
+	"github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+)
+
+const poolPath = "/latest/pools/"
+
+// ListPools and return them as obj
+func ListPools() (*v1alpha1.StoragePoolList, error) {
+
```
Comment (sonasingh46) [inline pkg/client/mapiserver/pool.go]: Extra blank line..

File: cmd/maya-apiserver/app/server/http.go [hunk 18]
```diff
@@ -136,6 +147,14 @@ var (
 		},
 		[]string{"code", "method"},
 	)
+
+	latestOpenEBSPoolRequestCounter = prometheus.NewCounterVec(
```
Comment (sonasingh46) [inline cmd/maya-apiserver/app/server/http.go]: A comment here would help.

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+	switch req.Method {
+	case "GET":
+		return poolOp.httpGet()
+	default:
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+}
+
+func (p *poolAPIOpsV1alpha1) httpGet() (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (sonasingh46) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: Ah! Above is fine. Ignore it. It is an http thing. I took the function as a util function.

File: pkg/install/v1alpha1/storage_pool.go (hunk 1 above)
Comment (ashishranjan738) [reply pkg/install/v1alpha1/storage_pool.go]: removed this as read is not included in this PR.

File: pkg/install/v1alpha1/storage_pool.go (hunk 2 above)
Comment (ashishranjan738) [reply pkg/install/v1alpha1/storage_pool.go]: removed this as read is not included in this PR.

File: pkg/storagepool/storagepool.go (hunk 3 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: done

File: pkg/storagepool/storagepool.go (hunk 4 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: done

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go (hunk 7 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: other endpoints are not having it should we add it ?

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go (hunk 8 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: fixed

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+type poolAPIOpsV1alpha1 struct {
+	req  *http.Request
+	resp http.ResponseWriter
+}
+
+// volumeV1alpha1SpecificRequest is a http handler to handle HTTP
+// requests to a OpenEBS volume.
+func (s *HTTPServer) poolV1alpha1SpecificRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	glog.Infof("cas template based pool request was received: method '%s'", req.Method)
```
[the remark this answers is not in the document]
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: done

File: pkg/storagepool/storagepool.go (hunk 9 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: fixed

File: pkg/storagepool/storagepool.go (hunk 11 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: fixed

File: pkg/storagepool/storagepool.go (hunk 12 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: fixed

File: pkg/storagepool/storagepool.go (hunk 13 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: fixed

File: pkg/storagepool/storagepool.go (hunk 14 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: fixed

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go (hunk 7 above)
Comment (sonasingh46) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: At least add for yours..
Others we can take it later.

File: pkg/storagepool/storagepool.go (hunk 15 above)
Comment (ashishranjan738) [reply pkg/storagepool/storagepool.go]: added

File: cmd/maya-apiserver/app/server/http.go (hunk 18 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/http.go]: added

File: cmd/maya-apiserver/app/server/http.go (hunk 5 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/http.go]: done

File: cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go (hunk 7 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/pool_endpoint_v1alpha1.go]: done

File: cmd/maya-apiserver/app/server/http.go (hunk 6 above)
Comment (ashishranjan738) [reply cmd/maya-apiserver/app/server/http.go]: done

**Comment by ashishranjan738** [comment]
closing this in favour of  https://github.com/openebs/maya/pull/805.

**Comment by ashishranjan738** [comment]
closing in favor of PR https://github.com/openebs/maya/pull/805.

**Pull request closed without merging**